### PR TITLE
Modify BoundsSan to improve debuggability

### PR DESCRIPTION
**Pull request by oskarwirga**
Context
BoundsSanitizer is a mitigation that is part of UBSAN. It can be enabled in "trap" mode to crash on OOB array accesses.

Problem
BoundsSan has zero false positives meaning every crash is a OOB array access, unfortunately optimizations cause these crashes in production builds to be a bit useless because we only know which function is crashing but not which line of code.

Godbolt example of the optimization: https://godbolt.org/z/6qjax9z1b

This Diff
I wanted to provide a way to know exactly which LOC is responsible for the crash. What we do here is use the size of the basic block as an iterator to an immediate value for the ubsan trap.

Previous discussion: https://reviews.llvm.org/D148654

### CODE REVIEW

**Comment by oskarwirga** [comment]
CC: @vitalybuka I addressed some but not all comments pending further clarification :)

File: llvm/lib/Transforms/Instrumentation/BoundsChecking.cpp [hunk 1]
```diff
@@ -180,19 +183,27 @@ static bool addBoundsChecking(Function &F, TargetLibraryInfo &TLI,
...
 
-    auto *F = Intrinsic::getDeclaration(Fn->getParent(), Intrinsic::trap);
-    CallInst *TrapCall = IRB.CreateCall(F, {});
+    Intrinsic::ID IntrID = DebugTrapBB ? Intrinsic::ubsantrap : Intrinsic::trap;
+    auto *F = Intrinsic::getDeclaration(Fn->getParent(), IntrID);
+
+    CallInst *TrapCall;
+    if (DebugTrapBB) {
+      TrapCall =
+          IRB.CreateCall(F, ConstantInt::get(IRB.getInt8Ty(), Fn->size()));
```
Comment (vitalybuka) [inline llvm/lib/Transforms/Instrumentation/BoundsChecking.cpp:202]: >> (TrapBB->getParent()->size() * 0x10000 + CheckHandlerID)

> Why * 0x10000 ?

anything that above CheckHandlerID so lower two bytes are still CheckHandlerID
and upper bytes you counter 

but i missed that it's i8
So keep your version

File: llvm/test/MC/AArch64/local-bounds-single-trap.ll [hunk 2]
```diff
@@ -0,0 +1,83 @@
+; RUN: llc -O3 -mtriple arm64-linux -filetype asm -o - %s | FileCheck %s -check-prefix CHECK-ASM
+; What this test does is check that even with nomerge, the functions still get merged in
+; compiled code as the ubsantrap call gets lowered to a single instruction: brk.
```
Comment (vitalybuka) [inline llvm/test/MC/AArch64/local-bounds-single-trap.ll:3]: you don't change MC, so I don't think we need the test here

However we need a test in llvm-project/llvm/test/Instrumentation/BoundsChecking/

**Review by vitalybuka — CHANGES_REQUESTED** [review]
the code LGTM,
but llvm-project/llvm/test/Instrumentation/BoundsChecking/ test is needed

Maybe just extend llvm-project/llvm/test/Instrumentation/BoundsChecking/simple.ll

File: llvm/test/MC/AArch64/local-bounds-single-trap.ll (hunk 2 above)
Comment (oskarwirga) [reply llvm/test/MC/AArch64/local-bounds-single-trap.ll:3]: I understood your comment on the Phabricator diff as a request for this test:

> can you please create a test where bounds-checking-single-trap=0 and setCannotMerge produce invalid result.

I will also add a test in `llvm/test/Instrumentation/BoundsChecking/`

**Review by vitalybuka — APPROVED** [review]

**Comment by oskarwirga** [comment]
@vitalybuka Thank you for reviewing! Can you merge this? I don't have write access (yet!)

**Comment by vitalybuka** [comment]
oskarwirga@ I'd like to use this feature but without counter, preserving ubsan IDs

Also I think in the current the counter has limited use: in optimized code, after inlining, it will have a lot of same ids, like 0, 1 from different functions.

So I propose to undo that part of the patch. WDYT?

**Comment by oskarwirga** [comment]
> @oskarwirga I'd like to use this feature but without counter, preserving ubsan IDs
> 
> Also I think in the current the counter has limited use: in optimized code, after inlining, it will have a lot of same ids, like 0, 1 from different functions.
> 
> So I propose to undo that part of the patch. WDYT?

So long as the debuggability is preserved so that it is possible to attribute crashes to specific lines of code, I am fine! Do you have an alternate solution?

**Comment by vitalybuka** [comment]
> So long as the debuggability is preserved so that it is possible to attribute crashes to specific lines of code, I am fine! Do you have an alternate solution?

In  IR these traps are not de-duplicated with -ubsan-unique-traps
It happens later, in LLVM backend, it needs to be fixed.
Another alternative just to use function call, e.g using https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#minimal-runtime.

**Pull request merged**